url/url.h

```diff
@@ -0,0 +1,40 @@
+// SPDX-FileCopyrightText: 2023 David Zero <zero-one@zer0-one.net>
+//
+// SPDX-License-Identifier: BSD-2-Clause
+
+#ifndef URL_URL_H_
+#define URL_URL_H_
+
+#include <cstdint>
+#include <optional>
+#include <span>
+#include <string>
+#include <variant>
+
+namespace url {
+
+enum class HostType { DNS_DOMAIN, IP4_ADDR, IP6_ADDR, OPAQUE, EMPTY };
+
+struct Host {
+    HostType type;
+
+    std::variant<std::string, std::uint32_t, std::span<std::uint16_t, 8>> data;
```

Comment:
`std::span` -> `std::array`.
A data structure that owns data feels like it makes more sense here since relying on the user to keep the Ipv6-addr alive separately from this struct isn't very intuitive.

Reply:
ah, that's right. whoops